kani-compiler/src/kani_middle/resolve.rs

```diff
@@ -258,41 +259,139 @@ fn resolve_relative(
+    }
+}
+
+/// Resolves a path by exploring a non-glob use statement.
+fn resolve_in_use(tcx: TyCtxt, use_path: &rustc_hir::Path, segments: Segments) -> Option<DefId> {
+    if let Res::Def(def_kind, def_id) = use_path.res {
+        tracing::debug!(
+            "Resolving `{}` via `use` import of `{}`",
+            segments_to_string(&segments),
+            tcx.def_path_str(def_id)
+        );
+        match def_kind {
+            DefKind::Fn => {
+                if segments.is_empty() {
+                    tracing::debug!(
+                        "Resolved `{}` to a function via `use` import of `{}`",
+                        segments_to_string(&segments),
+                        tcx.def_path_str(def_id)
+                    );
+                    return Some(def_id);
+                }
+            }
+            DefKind::Mod => return resolve_in_module(tcx, def_id, segments),
+            DefKind::Struct | DefKind::Enum | DefKind::Union => {
+                if segments.len() == 1 {
+                    return resolve_in_type(tcx, def_id, &segments[0]);
+                }
+            }
+            _ => (),
+        }
+    }
+    tracing::debug!("Unable to resolve `{}` via `use` import", segments_to_string(&segments));
     None
 }
 
-/// Resolves a name in an `impl` block.
-fn resolve_in_impl(tcx: TyCtxt, impl_id: DefId, name: &str) -> Option<DefId> {
-    tracing::debug!("Resolving `{name}` in impl block `{}`", tcx.def_path_str(impl_id));
-    for assoc_item in tcx.associated_item_def_ids(impl_id) {
-        let item_path = tcx.def_path_str(*assoc_item);
-        let last = item_path.split("::").last().unwrap();
-        if last == name {
-            tracing::debug!("Resolved `{name}` in impl block `{}`", tcx.def_path_str(impl_id));
-            return Some(*assoc_item);
+/// Resolves a path by exploring glob use statements.
+fn resolve_in_glob_uses(
+    tcx: TyCtxt,
+    current_module: LocalDefId,
+    glob_imports: Vec<&rustc_hir::Path>,
+    segments: &Segments,
+) -> Option<DefId> {
+    let glob_resolves = glob_imports
+        .iter()
+        .filter_map(|use_path| {
+            let span = tracing::span!(tracing::Level::DEBUG, "glob_resolution");
+            let _enter = span.enter();
+            resolve_in_glob_use(tcx, use_path, segments.clone())
+        })
+        .collect::<Vec<_>>();
+    if glob_resolves.len() == 1 {
+        return glob_resolves.first().copied();
+    }
+    if glob_resolves.len() > 1 {
+        // Raise an error if it's ambiguous which glob import a function comes
+        // from. rustc will also raise an error in this case if the ambiguous
+        // function is present in code (and not just as an attribute argument).
+        // TODO: We should make this consistent with error handling for other
+        // cases (see <https://github.com/model-checking/kani/issues/2013>).
+        let location = module_to_string(tcx, current_module);
+        let mut msg = format!(
```

Comment:
We currently don't have the span available anywhere in this module, and at this point in the code we don't even have the full name of the original function we're trying to resolve. We could potentially pass this information around all the different resolution functions so that we can produce helpful error messages here. However, I think it would be cleaner to change the resolution functions to return a `Result` and then leave it up to the client of this module to produce error messages.

Reply:
You could wrap these functions into a type that holds that information so you can generate more helpful error messages. Not on this PR though. :)

Reply:
Okay, sounds like we're holding off on making changes here until later. I think it probably makes most sense to go with the approach of returning a `Result`, so that this `resolve` module does not become specialized to stubbing, but can also be used in other parts of Kani.

Reply:
BTW, my suggestion is to keep it independent from stubbing. I'm OK with returning a result too, and letting the user to decide whether the failure to resolve is an error or not.

My main point is that the message should give more context on the exact path that we are trying to resolve and in which level it has failed.